builtin/provisioners/chef-client/resource_provisioner.go

```diff
@@ -0,0 +1,402 @@
+package chefclient
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path"
+	"regexp"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/hashicorp/terraform/communicator"
+	"github.com/hashicorp/terraform/communicator/remote"
+	"github.com/hashicorp/terraform/terraform"
+	"github.com/mitchellh/go-linereader"
+	"github.com/mitchellh/mapstructure"
+)
+
+const (
+	firstBoot      = "first-boot.json"
+	logfileDir     = "logfiles"
+	linuxConfDir   = "/etc/chef"
+	windowsConfDir = "C:/chef"
+)
+
+const clientConf = `
+log_location            STDOUT
+chef_server_url         "{{ .ServerURL }}"
+validation_client_name  "{{ .ValidationClientName }}"
+node_name               "{{ .NodeName }}"
+
+{{ if .HTTPProxy }}
+http_proxy          "{{ .HTTPProxy }}"
+ENV['http_proxy'] = "{{ .HTTPProxy }}"
+ENV['HTTP_PROXY'] = "{{ .HTTPProxy }}"
+{{ end }}
+
+{{ if .HTTPSProxy }}
+https_proxy          "{{ .HTTPSProxy }}"
+ENV['https_proxy'] = "{{ .HTTPSProxy }}"
+ENV['HTTPS_PROXY'] = "{{ .HTTPSProxy }}"
+{{ end }}
+
+{{ if .NOProxy }}no_proxy "{{ join .NOProxy "," }}"{{ end }}
+{{ if .SSLVerifyMode }}ssl_verify_mode {{ .SSLVerifyMode }}{{ end }}
+`
```

Comment:
This looks great. How do you feel about also allowing the user to specify a local path to their own config? Maybe you did that already and I haven't gotten that far in the review :smile: 

In my experience, it's very difficult to get all the options people want (especially once plugins are involved), so it might be good to have a "basic" template here, and then for more complex use cases, users can use their own.